pallets/providers/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+            // 3b. Any value proposition checks??? TODO: Ask
+            // 4. Check that the data to be stored is greater than the minimum required by the runtime. TODO: Ask if this applies to MSPs
+            // 5. Calculate how much deposit will the signer have to pay using the amount of data it wants to store
+            // 6. Check that the signer has enough funds to pay the deposit
+            // 7. Hold the deposit from the signer
+            // 8. Update the MSP storage to add the signer as an MSP
+            // 9. Increment the storage that holds total amount of SPs currently in the system
+            // 10. Emit an event confirming that the registration of the MSP has been successful
+
+            // Check that the extrinsic was signed and get the signer.
+            // This function will return an error if the extrinsic is not signed.
+            // https://docs.substrate.io/v3/runtime/origins
+            let who = ensure_signed(origin)?;
+
+            let msp_info = MainStorageProvider {
+                total_data,
+                data_used: StorageData::<T>::default(),
+                multiaddress,
+                value_prop,
+            };
+            // Update storage.
+            Self::do_msp_sign_up(&who, &msp_info)?;
+
+            // Emit an event.
+            Self::deposit_event(Event::<T>::MspSignUpSuccess {
+                who,
+                info: msp_info,
+            });
+            // Return a successful DispatchResultWithPostInfo
+            Ok(().into())
+        }
+
+        /// A dispatchable function that allows users to sign up as a backup storage provider.
+        #[pallet::call_index(1)]
+        #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+        pub fn bsp_sign_up(
+            origin: OriginFor<T>,
+            total_data: StorageData<T>,
+            multiaddress: MultiAddress<T>,
+        ) -> DispatchResultWithPostInfo {
+            // todo!("Logic to sign up a BSP.");
+            // This extrinsic should:
+            // 1. Check that the extrinsic was signed and get the signer.
+            // 2. Check that, by adding this new BSP, we won't exceed the max amount of BSPs allowed
+            // 3. Check that the signer is not already registered as either a MSP or BSP
+            // 4. Check that the multiaddress is valid (size and any other relevant checks)
+            // 5. Check that the data to be stored is greater than the minimum required by the runtime
+            // 6. Calculate how much deposit will the signer have to pay using the amount of data it wants to store
+            // 7. Check that the signer has enough funds to pay the deposit
+            // 8. Hold the deposit from the signer
+            // 9. Update the BSP storage to add the signer as an BSP
+            // 10. Update the total capacity of all BSPs, adding the new capacity (factoring by redundancy)
```

Comment:
I don't think you need to factor in redundancy here. If we have 10 BSPs with 10GB each of capcacity, we just say that the total capacity is 100GB. Then as this starts to get used, then we factor in redundancy in the capacity used. For example, a 1GB file, with 6x redundancy would be using 6GB of the BSP's total capacity.